File: play-with-lambda-expressions/play_with_lambda/ports.py
"""Abstract protocols for IO boundaries."""

from __future__ import annotations

from typing import Protocol

from .domain import Term


class TermGenerator(Protocol):
    """Generates lambda terms (e.g., randomly)."""

    def generate(self, max_depth: int) -> Term:
        """Produce a term with AST depth at most max_depth."""
        ...


class NameRegistry(Protocol):
    """Session-scoped named term storage."""

    def bind(self, name: str, term: Term) -> None:
        """Associate a name with a term (overwrites if exists)."""
        ...

    def lookup(self, name: str) -> Term | None:
        """Return the term bound to name, or None."""
        ...

    def all_names(self) -> dict[str, Term]:
        """Return all name-term bindings."""
        ...


class UserInterface(Protocol):
    """REPL interaction abstraction."""

    def read_input(self) -> str | None:
        """Read one line from the user. Return None on EOF."""
        ...

    def display(self, text: str) -> None:
        """Show output to the user."""
        ...

    def display_error(self, text: str) -> None:
        """Show an error message to the user."""
        ...

File: play-with-lambda-expressions/src/lambda_calculus/src/ports.rs
//! Abstract traits for IO boundaries (ports).

use std::collections::HashMap;
use crate::domain::Term;

/// Generates lambda terms (e.g., randomly).
pub trait TermGenerator {
    fn generate(&mut self, max_depth: usize) -> Term;
}

/// Session-scoped named term storage.
pub trait NameRegistry {
    fn bind(&mut self, name: &str, term: Term);
    fn lookup(&self, name: &str) -> Option<&Term>;
    fn all_names(&self) -> HashMap<String, Term>;
}

/// REPL interaction abstraction.
pub trait UserInterface {
    fn read_input(&mut self) -> Option<String>;
    fn display(&mut self, text: &str);
    fn display_error(&mut self, text: &str);
}
